src/VisualStudio/IntegrationTests/CSharp/CSharpAutomaticBraceCompletion.cs

```diff
@@ -0,0 +1,146 @@
+            _project = _solution.AddProject("TestProj", ProjectTemplate.ClassLibrary, ProjectLanguage.CSharp);
+
+            _workspace = _visualStudio.Instance.Workspace;
+            _workspace.UseSuggestionMode = false;
+
+            _editorWindow = _visualStudio.Instance.EditorWindow;
+        }
+
+        public void Dispose()
+        {
+            _editorWindow.TypeText("\u001B");
+            _visualStudio.Dispose();
+        }
+
+        [Fact]
+        public void BracesInsertionAndTabCompleting()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+            VerifyEditorTextBeforeAndAfterCursor("        if (true) { ", "}");
+
+            _editorWindow.TypeText("\t");
+            VerifyEditorTextBeforeCursor("        if (true) { }");
+        }
+
+        [Fact]
+        public void BracesOvertyping()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+
+            _editorWindow.TypeText("}");
+            VerifyEditorTextBeforeCursor("        if (true) { }");
+        }
+
+        [Fact]
+        public void BracesOnReturnNoFormattingOnlyIndentationBeforeCloseBrace()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+
+            _editorWindow.TypeText("\rvar a = 1;");
+            VerifyEditorTextBeforeCursor("            var a = 1;");
+        }
+
+        [Fact]
+        public void BracesOnReturnOvertypingTheClosingBrace()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+
+            _editorWindow.TypeText("\rvar a = 1;}");
+            VerifyEditorTextBeforeCursor("        }");
+
+            VerifyEditorTextContains(@"if (true)
+        {
+            var a = 1;
+        }
+");
+        }
+
+        [Fact]
+        public void BracesOnReturnWithNonWhitespaceSpanInside() // DevDiv 653540
+        {
+            _editorWindow.Text = string.Empty;
+            _editorWindow.TypeText("class A { int i;\r");
+
+            VerifyEditorTextAfterCursor("}");
+            VerifyEditorTextContains(@"class A { int i;
+}");
+        }
+
+        private void VerifyEditorTextAfterCursor(string expectedText) => VerifyEditorTextBeforeAndAfterCursor(string.Empty, expectedText);
+
+        private void VerifyEditorTextBeforeCursor(string expectedText) => VerifyEditorTextBeforeAndAfterCursor(expectedText, string.Empty);
+
+        private void VerifyEditorTextContains(string expectedText)
```

Comment:
I would like to see the xUnit primitives being used directly; otherwise it just feels like we're porting our actions in a strange way. Put another way, this adds a layer of indirection which doesn't simplify anything and just increases concept count.